### ISSUE TITLE
Optimizer and traning frequency

### ISSUE THREAD

**Issue by xfdywy** [ShangtongZhang/DeepRL#14]
The optimizer in origin Deepmind DQN paper is not the traditional RMSprop,
you can refer to here [https://github.com/spragunr/deep_q_rl/blob/master/deep_q_rl/updates.py](url)
and section 6.2 in this paper [https://arxiv.org/pdf/1711.07478.pdf](url)

Also, often we update the Q network after 4 step or something like  'update frequency = 4', I cannot find such implementation in your DQN code.

Please correct me if I am wrong.

**Comment by ShangtongZhang** [comment]
Yes it's different. I remember you can set `centered=True` in PyTorch to make them the same. But I'm not sure.

Do you mean frame skip? It exists in atari wrapper.

**Comment by xfdywy** [comment]
Not only frame skip. I mean during the training process, you don't have to calculate gradient and then do bp every step. 

By the way, you don't implement any clip for gradient?

**Comment by ShangtongZhang** [comment]
>  I mean during the training process, you don't have to calculate gradient and then do bp every step.

I don't know this trick, do you know where it comes from?

> By the way, you don't implement any clip for gradient?

In my understanding gradient clip is only necessary in multi-process setting

**Comment by xfdywy** [comment]
update_freq=4  defined here
[ https://github.com/kuz/DeepMind-Atari-Deep-Q-Learner/blob/1d52e98152fccd968d63ca67e33f5cbf3e83a878/run_gpu#L14](url)


and used here
[https://github.com/kuz/DeepMind-Atari-Deep-Q-Learner/blob/1d52e98152fccd968d63ca67e33f5cbf3e83a878/dqn/NeuralQLearner.lua#L341](url)


clip_delta=1
[https://github.com/kuz/DeepMind-Atari-Deep-Q-Learner/blob/1d52e98152fccd968d63ca67e33f5cbf3e83a878/run_gpu#L31](url)

**Comment by ShangtongZhang** [comment]
Thanks, I haven't read the original lua code before.
It's interesting to know that tricks.

**Comment by xfdywy** [comment]
Another issue................. In deepmind code, gradient clip  is really "clip"

 if self.clip_delta then
        delta[delta:ge(self.clip_delta)] = self.clip_delta
        delta[delta:le(-self.clip_delta)] = -self.clip_delta
    end

But you use nn.utils.clip.grad_norm, which is more likely to do "normalization" .....
http://pytorch.org/docs/master/_modules/torch/nn/utils/clip_grad.html

**Comment by ShangtongZhang** [comment]
Yes .. I know that, just to conform with pytorch.

**Comment by xfdywy** [comment]
Ok, thanks! So many details................
